File: app/services/question_service.py
"""Service for question CRUD operations."""

from typing import List, Optional, Tuple, Dict, Any
from sqlalchemy.orm import Session
from sqlalchemy import or_, and_
from app.models import Question, QuestionUsage
from app.schemas import QuestionCreate, QuestionUpdate, QuestionUsageCreate
from app.services.similarity_service import (
    generate_hash,
    generate_embedding,
    check_similarity,
    check_exact_duplicate,
)


class DuplicateQuestionError(ValueError):
    """Custom exception for duplicate or similar questions."""

    def __init__(self, message: str, detail: Optional[Dict[str, Any]] = None):
        super().__init__(message)
        self.detail = detail or {"message": message}

    def to_detail(self) -> Dict[str, Any]:
        return {"message": str(self), **self.detail}


def create_question(db: Session, question: QuestionCreate) -> Question:
    """
    Create a new question in the database.
    Checks for duplicates within the same college before creating.
    If an exact duplicate exists for the same college, increment usage_count
    instead of creating a new record.

    Args:
        db (Session): Database session.
        question (QuestionCreate): Question data.

    Returns:
        Question: Created question object.

    Raises:
        ValueError: If duplicate question found in the same college.
    """
    # Check for existing exact matches (any college)
    question_hash = generate_hash(question.question_text)
    existing_exact_all = (
        db.query(Question)
        .filter(
            Question.question_hash == question_hash,
            Question.status == "Active",
        )
        .all()
    )

    # If an exact duplicate exists in the same college, raise an error
    existing_same_college = next(
        (q for q in existing_exact_all if q.college == question.college), None
    )
    if existing_same_college:
        raise DuplicateQuestionError(
            message=(
                f"Question already exists in college '{question.college}'. "
                f"Please use the existing question (ID: {existing_same_college.id}) or modify your question text."
            ),
            detail={
                "code": "DUPLICATE_QUESTION_SAME_COLLEGE",
                "college": question.college,
                "existing_question_id": existing_same_college.id,
            },
        )

    # If exact matches exist in other colleges, synchronize usage_count across all
    if existing_exact_all:
        # New total usage is number of exact matches (including the new one)
        new_total_usage = len(existing_exact_all) + 1
        # Update existing matches to reflect new total usage
        for q in existing_exact_all:
            q.usage_count = new_total_usage
        db.flush()  # stage updates before adding new question
        # If exact duplicate exists in another college, allow adding to new college
        # Skip similarity check since we're allowing cross-college duplicates
        # Only check similarity if no exact duplicates exist anywhere
    else:
        # Check for similar questions within the same college (only if no exact duplicate exists)
        is_duplicate, exact_match, similar_questions = check_similarity(
            db, question.question_text, college=question.college
        )
        if is_duplicate:
            similar_count = len(similar_questions)
            similar_ids = [q.id for q, _ in similar_questions[:5]]  # Show first 5 IDs
            raise DuplicateQuestionError(
                message=(
                    f"Similar question(s) found in college '{question.college}'. "
                    "Please review existing questions or modify your question text."
                ),
                detail={
                    "code": "SIMILAR_QUESTION",
                    "college": question.college,
                    "similar_count": similar_count,
                    "similar_question_ids": similar_ids,
                },
            )
    
    # Generate hash and embedding
    embedding = generate_embedding(question.question_text)
    
    db_question = Question(
        question_text=question.question_text,
        question_hash=question_hash,
        subject=question.subject,
        topic=question.topic,
        difficulty_level=question.difficulty_level,
        marks=question.marks,
        exam_type=question.exam_type,
        college=question.college,
        embedding=embedding,
        status="Active",
        usage_count=len(existing_exact_all) + 1 if existing_exact_all else 1,
    )
    
    db.add(db_question)
    db.commit()
    db.refresh(db_question)
    
    return db_question


def get_question(db: Session, question_id: int) -> Optional[Question]:
    """
    Get a question by ID.

    Args:
        db (Session): Database session.
        question_id (int): Question ID.

    Returns:
        Optional[Question]: Question if found, None otherwise.
    """
    return db.query(Question).filter(Question.id == question_id).first()


def get_all_questions(
    db: Session,
    skip: int = 0,
    limit: int = 100,
    subject: Optional[str] = None,
    exam_type: Optional[str] = None,
    college: Optional[str] = None,
    status: Optional[str] = None,
    search_text: Optional[str] = None,
) -> Tuple[List[Question], int]:
    """
    Get all questions with optional filters.

    Args:
        db (Session): Database session.
        skip (int): Number of records to skip.
        limit (int): Maximum number of records to return.
        subject (Optional[str]): Filter by subject.
        exam_type (Optional[str]): Filter by exam type.
        college (Optional[str]): Filter by college.
        status (Optional[str]): Filter by status.
        search_text (Optional[str]): Search in question text.

    Returns:
        Tuple[List[Question], int]: List of questions and total count.
    """
    query = db.query(Question)
    
    # Apply filters
    if subject:
        query = query.filter(Question.subject.ilike(f"%{subject}%"))
    if exam_type:
        query = query.filter(Question.exam_type == exam_type)
    if college:
        query = query.filter(Question.college.ilike(f"%{college}%"))
    if status:
        query = query.filter(Question.status == status)
    if search_text:
        query = query.filter(Question.question_text.ilike(f"%{search_text}%"))
    
    # Get total count
    total = query.count()
    
    # Apply pagination
    questions = query.order_by(Question.created_date.desc()).offset(skip).limit(limit).all()
    
    return questions, total


def update_question(db: Session, question_id: int, question_update: QuestionUpdate) -> Optional[Question]:
    """
    Update a question.

    Args:
        db (Session): Database session.
        question_id (int): Question ID.
        question_update (QuestionUpdate): Updated question data.

    Returns:
        Optional[Question]: Updated question if found, None otherwise.
    """
    db_question = get_question(db, question_id)
    if not db_question:
        return None
    
    # Update fields
    update_data = question_update.model_dump(exclude_unset=True)
    
    # If question text is updated, regenerate hash and embedding
    if "question_text" in update_data:
        db_question.question_hash = generate_hash(update_data["question_text"])
        db_question.embedding = generate_embedding(update_data["question_text"])
    
    for field, value in update_data.items():
        setattr(db_question, field, value)
    
    db.commit()
    db.refresh(db_question)
    
    return db_question


def delete_question(db: Session, question_id: int) -> bool:
    """
    Permanently delete a question from the database.

    Args:
        db (Session): Database session.
        question_id (int): Question ID.

    Returns:
        bool: True if question was found and deleted, False otherwise.
    """
    try:
        db_question = get_question(db, question_id)
        if not db_question:
            return False
        
        # Permanently delete the question
        # Note: This will also delete related usage history due to cascade="all, delete-orphan"
        db.delete(db_question)
        db.commit()
        
        return True
    except Exception as e:
        db.rollback()
        raise e


def record_question_usage(db: Session, question_id: int, usage: QuestionUsageCreate) -> QuestionUsage:
    """
    Record that a question has been used in an exam.

    Args:
        db (Session): Database session.
        question_id (int): Question ID.
        usage (QuestionUsageCreate): Usage information.

    Returns:
        QuestionUsage: Created usage record.
    """
    # Get question and update usage count
    db_question = get_question(db, question_id)
    if not db_question:
        raise ValueError(f"Question with ID {question_id} not found")
    
    # Create usage record
    db_usage = QuestionUsage(
        question_id=question_id,
        exam_name=usage.exam_name,
        exam_type=usage.exam_type,
        academic_year=usage.academic_year,
        college=usage.college,
    )
    
    db.add(db_usage)
    
    # Update question usage statistics
    db_question.usage_count += 1
    from datetime import datetime
    db_question.last_used_date = datetime.utcnow()
    
    db.commit()
    db.refresh(db_usage)
    
    return db_usage


def get_question_usage_history(db: Session, question_id: int) -> List[QuestionUsage]:
    """
    Get usage history for a question.

    Args:
        db (Session): Database session.
        question_id (int): Question ID.

    Returns:
        List[QuestionUsage]: List of usage records.
    """
    return (
        db.query(QuestionUsage)
        .filter(QuestionUsage.question_id == question_id)
        .order_by(QuestionUsage.date_used.desc())
        .all()
    )


def check_question_duplicate(db: Session, question_text: str, exclude_id: Optional[int] = None, college: Optional[str] = None) -> Tuple[bool, bool, List[Tuple[Question, float]]]:
    """
    Check if a question is duplicate or similar to existing questions.
    If college is provided, only checks within that college.

    Args:
        db (Session): Database session.
        question_text (str): Question text to check.
        exclude_id (Optional[int]): Question ID to exclude from results.
        college (Optional[str]): College name to filter by. If provided, only checks duplicates in this college.

    Returns:
        Tuple[bool, bool, List[Tuple[Question, float]]]:
            - is_duplicate: True if exact or similar match found
            - exact_match: True if exact match found
            - similar_questions: List of (question, similarity_score) tuples
    """
    return check_similarity(db, question_text, exclude_id=exclude_id, college=college)

File: backend-express/services/questionService.js
/**
 * Question service - handles business logic for questions
 */

const storage = require('./storage');
const similarity = require('./similarity');

class DuplicateQuestionError extends Error {
  constructor(message, detail = {}) {
    super(message);
    this.detail = { message, ...detail };
  }
}

/**
 * Create a new question
 */
async function createQuestion(questionData) {
  const questions = await storage.getAllQuestions();
  const questionHash = similarity.generateHash(questionData.question_text);
  
  // Check for existing exact matches (any college)
  const existingExactAll = questions.filter(
    q => q.question_hash === questionHash && q.status === 'Active'
  );
  
  // If exact duplicate exists in same college, raise error
  const existingSameCollege = existingExactAll.find(q => q.college === questionData.college);
  if (existingSameCollege) {
    throw new DuplicateQuestionError('Question already exists', {
      code: 'DUPLICATE_QUESTION_SAME_COLLEGE',
      college: questionData.college,
      existing_question_id: existingSameCollege.id
    });
  }
  
  // If exact matches exist in other colleges, synchronize usage_count
  if (existingExactAll.length > 0) {
    const newTotalUsage = existingExactAll.length + 1;
    for (const q of existingExactAll) {
      await storage.updateQuestion(q.id, { usage_count: newTotalUsage });
    }
  } else {
    // Check for similar questions within the same college
    const { isDuplicate, similarQuestions } = similarity.checkSimilarity(
      questions,
      questionData.question_text,
      null,
      questionData.college
    );
    
    if (isDuplicate && similarQuestions.length > 0) {
      const similarCount = similarQuestions.length;
      const similarIds = similarQuestions.slice(0, 5).map(([q]) => q.id);
      throw new DuplicateQuestionError(
        `Similar question(s) found in college '${questionData.college}'. Please review existing questions or modify your question text.`,
        {
          code: 'SIMILAR_QUESTION',
          college: questionData.college,
          similar_count: similarCount,
          similar_question_ids: similarIds
        }
      );
    }
  }
  
  // Set usage_count
  const usageCount = existingExactAll.length > 0 ? existingExactAll.length + 1 : 1;
  
  // Create question
  const newQuestion = await storage.addQuestion({
    question_text: questionData.question_text,
    question_hash: questionHash,
    subject: questionData.subject,
    topic: questionData.topic,
    difficulty_level: questionData.difficulty_level,
    marks: questionData.marks,
    exam_type: questionData.exam_type,
    college: questionData.college,
    usage_count: usageCount,
    status: 'Active'
  });
  
  return newQuestion;
}

/**
 * Get question by ID
 */
async function getQuestion(id) {
  return await storage.getQuestionById(id);
}

/**
 * Get all questions with filters
 */
async function getAllQuestions(filters = {}) {
  let questions = await storage.getAllQuestions();
  
  // Apply filters - only filter if value is provided and not empty
  if (filters.subject && filters.subject.trim()) {
    const subjectLower = filters.subject.trim().toLowerCase();
    questions = questions.filter(q => q.subject && q.subject.toLowerCase().includes(subjectLower));
  }
  if (filters.exam_type && filters.exam_type.trim()) {
    questions = questions.filter(q => q.exam_type === filters.exam_type.trim());
  }
  if (filters.college && filters.college.trim()) {
    const collegeLower = filters.college.trim().toLowerCase();
    questions = questions.filter(q => q.college && q.college.toLowerCase().includes(collegeLower));
  }
  if (filters.status && filters.status.trim()) {
    questions = questions.filter(q => q.status === filters.status.trim());
  }
  if (filters.search_text && filters.search_text.trim()) {
    const searchLower = filters.search_text.trim().toLowerCase();
    questions = questions.filter(q => 
      q.question_text && q.question_text.toLowerCase().includes(searchLower)
    );
  }
  
  // Sort by created_date (descending)
  questions.sort((a, b) => {
    const dateA = new Date(a.created_date || 0);
    const dateB = new Date(b.created_date || 0);
    return dateB - dateA;
  });
  
  // Apply pagination
  const skip = parseInt(filters.skip) || 0;
  const limit = parseInt(filters.limit) || 100;
  const total = questions.length;
  const paginatedQuestions = questions.slice(skip, skip + limit);
  
  return {
    questions: paginatedQuestions,
    total
  };
}

/**
 * Update a question
 */
async function updateQuestion(id, updates) {
  const question = await storage.getQuestionById(id);
  if (!question) {
    return null;
  }
  
  const updateData = { ...updates };
  
  // If question text is updated, regenerate hash
  if (updates.question_text) {
    updateData.question_hash = similarity.generateHash(updates.question_text);
  }
  
  return await storage.updateQuestion(id, updateData);
}

/**
 * Delete a question
 */
async function deleteQuestion(id) {
  return await storage.deleteQuestion(id);
}

/**
 * Check question duplicate/similarity
 */
async function checkQuestionDuplicate(questionText, excludeId = null, college = null) {
  try {
    const questions = await storage.getAllQuestions();
    
    if (!questions || questions.length === 0) {
      return {
        isDuplicate: false,
        exactMatch: false,
        similarQuestions: []
      };
    }
    
    const result = similarity.checkSimilarity(
      questions,
      questionText,
      excludeId,
      college
    );
    
    return {
      isDuplicate: result.isDuplicate,
      exactMatch: result.exactMatch,
      similarQuestions: result.similarQuestions || []
    };
  } catch (error) {
    console.error('Error in checkQuestionDuplicate:', error);
    throw error;
  }
}

/**
 * Record question usage by question ID
 */
async function recordQuestionUsage(questionId, usageData) {
  const question = await storage.getQuestionById(questionId);
  if (!question) {
    throw new Error(`Question with ID ${questionId} not found`);
  }
  
  const usage = await storage.addUsageRecord({
    question_id: questionId,
    exam_name: usageData.exam_name,
    exam_type: usageData.exam_type,
    academic_year: usageData.academic_year,
    college: usageData.college
  });
  
  return usage;
}

/**
 * Record question usage by question text
 */
async function recordQuestionUsageByText(questionText, usageData) {
  const questions = await storage.getAllQuestions();
  const questionHash = similarity.generateHash(questionText);
  
  const matchingQuestions = questions.filter(
    q => q.question_hash === questionHash && q.status === 'Active'
  );
  
  if (matchingQuestions.length === 0) {
    throw new Error('Question not found for the given text');
  }
  
  // Record usage for the first matching question
  return await recordQuestionUsage(matchingQuestions[0].id, usageData);
}

/**
 * Get usage by question text
 */
async function getUsageByQuestionText(questionText) {
  return await storage.getUsageByQuestionText(questionText);
}

module.exports = {
  createQuestion,
  getQuestion,
  getAllQuestions,
  updateQuestion,
  deleteQuestion,
  checkQuestionDuplicate,
  recordQuestionUsage,
  recordQuestionUsageByText,
  getUsageByQuestionText,
  DuplicateQuestionError
};

